assets/js/wpjm-stats.js

```diff
@@ -4,7 +4,148 @@ import domReady from '@wordpress/dom-ready';
 import { createHooks } from '@wordpress/hooks';
 
 ( function () {
-	window.wpjmStatHooks = window.wpjmStatHooks || createHooks();
+	function debounce(func, delay) {
+		let timeoutId;
+
+		return function (...args) {
+			clearTimeout(timeoutId);
+
+			timeoutId = setTimeout(() => {
+				func(...args);
+			}, delay);
+		};
+	}
+
+	function createStatsQueue() {
+		const alreadySent = {};
+		let queue = [];
+
+		const logThem = debounce( function ( listingIds ) {
+			const stats = listingIds.map( function ( id ) {
+				alreadySent[id] = true;
+				return { name: 'job_listing_impressions', post_id: id };
+			});
+			console.log( 'sending stats:', stats );
+			// queue = [];
+			return wpjmLogStats( stats ).finally( function () {
+				queue = [];
+			});
+		}, 500 );
+
+		return {
+			queueListingImpressionStats: function ( listingIds ) {
+				listingIds.forEach( function (listingId ) {
+					if ( ! alreadySent[listingId] ) {
+						queue.push(listingId);
+					}
+				} );
+				logThem( queue );
+
+			}
+		};
+	}
+
+	function observeForVisibility( jobListingContainer, jobListingElement, visibleCallback, alreadyViewedListings ) {
+		const options = {
+			root: null,
+			rootMargin: "0px",
+			threshold: 1.0,
+		};
+
+		const observer = new IntersectionObserver(function ( entries ) {
+			entries.forEach(function ( entry ) {
+				if ( entry.isIntersecting && entry.intersectionRatio > 0.99 ) {
+					const node = entry.target;
+					if ( 1 === node.nodeType && node.classList.contains( 'job_listing' ) ) {
+						const nodeId = node.dataset.id;
+						if ( ! alreadyViewedListings[nodeId] ) {
+							alreadyViewedListings[nodeId] = true;
+							visibleCallback( node );
+						}
+					}
+					node.classList.add( 'viewed' );
+					observer.unobserve( node );
+				}
+			} );
+		}, options );
+
+		observer.observe( jobListingElement );
+	}
+
+	function waitForNextVisibleListing( listingVisibleCallback ) {
+		const jobListingsContainer = document.querySelector( 'ul.job_listings' );
+		const config = { childList: true };
+		const alreadyViewed = {};
+
+		const observer = new MutationObserver(function ( mutations ) {
+			mutations.forEach(function ( mutation ) {
+				mutation.addedNodes.forEach( function ( node ) {
+					if ( 1 === node.nodeType && node.classList.contains( 'job_listing' ) && ! node.classList.contains( 'viewed' ) ) {
```

Comment:
Would it make sense to make the selectors part of the config?

Reply:
If we end up reusing this for other things then maybe yes? Let's maybe do it in a consecutive PR.